### PR TITLE
Avoid duplicate serialization for TermsQueryBuilder

### PR DESCRIPTION
**Pull request by gf2121**
#67132 tried to cache all the results of serialization of `SearchSourceBuilder`, which is invasive. So this is a patch that only aims to optimizing `TermsQueryBuilder`. 

Here are some benchmark results (indices and queries are the same as #67132):

**No Param**
target | baseline | candidate
-- | -- | --
1 index (8 shards) | 83ms | 13ms
30 indices (240 shards) | 2935ms | 201ms

**set pre_filter_shard_size=240**

target | baseline | candidate
-- | -- | --
1 index (8 shards) | 81ms | 13ms
30 indices (240 shards) | 731ms | 104ms

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java [hunk 1]
```diff
@@ -342,6 +353,108 @@ public LongValuesSource rewrite(IndexSearcher searcher) throws IOException {
 
     }
 
+    private static final Set<Class<? extends Number>> INTEGER_TYPES = new HashSet<>(
```
Comment (gf2121) [inline server/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java:356]: #21776 made an effort to convert the List<?> in some special cases to help GC. But if we just serialize the list to a BytesRef, the convertion can be unnecessary. So i moved the method to the TermsSetQueryBuilder, which is the other query needs this method. I think TermsSetQueryBuilder's values do not need to be serialized because its terms number is limited to 1024.

**Comment by gf2121** [comment]
@jimczi Hi! Sorry to disturb. Could you please take a look at this when you are free?

File: server/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java (hunk 1 above)
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java:356]: I agree, we can consider this PR as another approach to handle large `terms` query but this is not needed for the `terms_set` query. Maybe that's a chance to cleanup this code a bit and remove the optimization entirely here if it's not needed. That can also be done in a follow up but in such case please add a TODO regarding the removal of this code.

**Review by jimczi — CHANGES_REQUESTED** [review]
Thanks @gf2121 , the approach makes sense to me. I left some comments to simplify the code a bit. The logic is a bit hard to follow currently, notably the back and forth between `BytesRef` and `String`. With this change we can make the conversion once since the list will be short lived (we can materialize values when the query is built).

File: server/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java [hunk 2]
```diff
@@ -63,18 +59,41 @@
...
+     * If we do not keep this, it could be expensive when receiving a request from
+     * lower version.
+     * We have to read the value list by {@link StreamInput#readGenericValue},
+     * serialize it into {@link #valueRef}, and then deserialize it again when
+     * {@link #doToQuery} called}.
+     * <p>
+     * ToDo: remove in 9.0.0
+     */
+    private final List<?> values;
+    private final boolean stale; //indicate if this comes from a lower version
```
[the remark this answers is not in the document]
Comment (gf2121) [reply server/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java]: @jimczi  Thank you for the review, I think this idea is really excellent! i've made some effort to optimize the code based on this idea in a new [commit](https://github.com/elastic/elasticsearch/pull/67223/commits/b374ec734abe868276853078679073bec602ed5c). Here are some notes about this modification
1. Using a `Iterable` to lazily load data from `BytesRef` is very good, but a `MappedFieldType#termsQuery` used to expect a `list` of terms, which means we still had to convert the `Iterable` to `List` to get a TermsQuery. This can make the lazily load meaningless, so i changed the method signature from a `List` to a `Collection`.
2. The reason that I did not convert `List` to `Iterable` but `Collection` is that we usually use a `List#size()` to pre-construct another Collection, e.g. a `LongHashSet` in `DateScriptFieldType`. This can help when we add data into the new collection. we can not get a `size` value from `Iterable`, but we can get it from `Collection`.

File: server/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java (hunk 1 above)
Comment (gf2121) [reply server/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java:357]: I'd like to fix this in a folow up. A TODO note has been added here.

File: server/src/main/java/org/elasticsearch/index/mapper/ConstantFieldType.java [hunk 3]
```diff
@@ -88,7 +88,7 @@ public final Query termQueryCaseInsensitive(Object value, QueryShardContext cont
     }
 
     @Override
-    public final Query termsQuery(List<?> values, QueryShardContext context) {
+    public final Query termsQuery(Collection<?> values, QueryShardContext context) {
```
Comment (gf2121) [inline server/src/main/java/org/elasticsearch/index/mapper/ConstantFieldType.java:91]: Here are some notes about this modification
1. Using `Iterable` to load data from `BytesRef` lazily is very good, but the `MappedFieldType#termsQuery` used to expect a `list` of terms, which means we still had to convert the `Iterable` to `List` to build a TermsQuery. This can make the lazily load meaningless, so i changed the method signature from a `List` to a `Collection`.
2. The reason that I did not convert `List` to `Iterable` but `Collection` is that we usually use a `List#size()` to pre-construct another Collection, e.g. a `LongHashSet` in `DateScriptFieldType`. This can help when we add data into the new collection. we can not get a `size` value from `Iterable`, but we can get it from `Collection`.

File: server/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java [hunk 4]
```diff
@@ -500,4 +399,249 @@ protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) {
...
+                    arrayList.add(o);
+                }
+                list = arrayList;
+            }
+            try (BytesStreamOutput output = new BytesStreamOutput()) {
+                if (convert) {
+                    list = list.stream().map(AbstractQueryBuilder::maybeConvertToBytesRef).collect(Collectors.toList());
+                }
+                output.writeGenericValue(list);
+                return output.bytes().toBytesRef();
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java:442]: Drive-by comment: this call to `BytesReference#toBytesRef()` is typically going to allocate a whole new large `byte[]` just to copy all the pages of the stream into it. That seems to be an unnecessary expense, I don't see a need for the result of this method to be contiguous in memory, so let's use `BytesReference` rather than `BytesRef` to avoid this copying.

File: server/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java [hunk 5]
```diff
@@ -342,6 +353,109 @@ public LongValuesSource rewrite(IndexSearcher searcher) throws IOException {
...
+                    }
+                    ++i;
+                }
+                final BytesReference bytes = bytesOut.bytes();
+                return new AbstractList<Object>() {
+                    @Override
+                    public Object get(int i) {
+                        final int startOffset = i == 0 ? 0 : endOffsets[i-1];
+                        final int endOffset = endOffsets[i];
+                        return bytes.slice(startOffset, endOffset - startOffset).toBytesRef();
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java:446]: Similarly here. Let's keep things in a `BytesReference`.

**Review by DaveCTurner — COMMENTED** [review]
I have no opinions on the big-picture changes here, but there's a couple of calls to `BytesReference#toBytesRef()` that I think shouldn't be there.

**Comment by DaveCTurner** [comment]
> @gf2121 gf2121 force-pushed the gf2121:terms_query_optmize  branch from 99c6fe5 to e054576 1 hour ago

@gf2121 please don't force-push to branches under review, it loses comments and context.

File: server/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java (hunk 1 above)
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java:357]: ++, please use `TODO` in capital letters for consistency.

**Comment by gf2121** [comment]
> > @gf2121 gf2121 force-pushed the gf2121:terms_query_optmize  branch from [99c6fe5](https://github.com/elastic/elasticsearch/commit/99c6fe501b18a7d61c5ad43ef02212922722ce1d) to [e054576](https://github.com/elastic/elasticsearch/commit/e054576de617839c8d0415fb09b21e8d80acc6ce) 1 hour ago
> 
> @gf2121 please don't force-push to branches under review, it loses comments and context.

OK!

File: server/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java (hunk 4 above)
Comment (gf2121) [reply server/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java]: @DaveCTurner Thanks for the review! fixed in [b2b9afd](https://github.com/elastic/elasticsearch/pull/67223/commits/b2b9afd5e81a9921abef3cbf25aff1d9083d87b0)

File: server/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java (hunk 5 above)
Comment (gf2121) [reply server/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java:446]: @DaveCTurner  This is just some code existed originally, and need to be remove in a follow up. so maybe we do not need to optimize this? :)

File: server/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java (hunk 1 above)
Comment (gf2121) [reply server/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java:357]: Thanks! fixed in [b2b9afd](https://github.com/elastic/elasticsearch/pull/67223/commits/b2b9afd5e81a9921abef3cbf25aff1d9083d87b0).

File: server/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java (hunk 5 above)
Comment (gf2121) [reply server/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java:446]: You can find more context [here](https://github.com/elastic/elasticsearch/pull/67223#discussion_r554591349)

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java:446]: Ok I see. These ones are small so will mostly avoid copying.

**Comment by gf2121** [comment]
@jimczi @DaveCTurner Hi! Sorry to disturb. I'm wonder if this PR is ready for a merge? Or just tell me if there are some more optimizations need for this :)

File: server/src/test/java/org/elasticsearch/index/query/TermsQueryBuilderTests.java [hunk 6]
```diff
@@ -345,4 +327,155 @@ protected QueryBuilder parseQuery(XContentParser parser) throws IOException {
...
+        {
+            TermsQueryBuilderBeforeV8 builder = new TermsQueryBuilderBeforeV8("foo", Arrays.asList(new BytesRef("a"), new BytesRef("b"), new BytesRef("c")), null);
+            TermsQueryBuilder copy = (TermsQueryBuilder) serializeAndDeserialize(Version.V_7_12_0, builder, Version.V_8_0_0, TermsQueryBuilder::new);
+            List<Object> values = copy.values();
+            assertEquals(Arrays.asList("a", "b", "c"), values);
+            assertTrue(copy.getValues() instanceof TermsQueryBuilder.ValuesBeforeV8);
+        }
+    }
+
+    public void testSerializationToLowerVersion() throws IOException {
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/index/query/TermsQueryBuilderTests.java:361]: It would be easier to maintain if the version are randomized. You can use VersionUtils#randomCompatibleVersion(random, Version.CURRENT) to pick compatible versions randomly.
Otherwise we need to change the checks on every backport/new branch.

File: server/src/test/java/org/elasticsearch/index/query/TermsQueryBuilderTests.java [hunk 7]
```diff
@@ -345,4 +327,155 @@ protected QueryBuilder parseQuery(XContentParser parser) throws IOException {
         return query;
     }
 
+    public void testSerializationFromLowerVersion() throws IOException {
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/index/query/TermsQueryBuilderTests.java:330]: That seems fragile since the serialization could change in lower version after this PR so we can diverge.
It's a good thing to test serialization in a unit test but it should be limited to the code that is implemented in this branch (without TermsQueryBuilderBeforeV8).

**Review by jimczi — CHANGES_REQUESTED** [review]
I left some comments regarding the implementation and the tests but we're getting close , thanks @gf2121 !

File: server/src/main/java/org/elasticsearch/index/mapper/ConstantFieldType.java (hunk 3 above)
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/index/mapper/ConstantFieldType.java:91]: ++ to move from a List to a Collection in `termsQuery`. That touches a lot of files but the change makes sense to me.

File: server/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java [hunk 8]
```diff
@@ -500,4 +399,255 @@ protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) {
 
         return this;
     }
+
+    public static abstract class Values extends AbstractCollection implements Writeable {
```
[the remark this answers is not in the document]
Comment (gf2121) [reply server/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java:403]: `TermsQueryBuilder` provides a public method `getValues()`, which returns a `Values`. It seems that `Values` is not used outside now because the `getValues()` is used in this pattern: 
```
return new TermsQueryBuilder(rewrittenFieldName, terms.getValues());
```
So maybe this class should be public since there is such a method? :)

File: server/src/test/java/org/elasticsearch/index/query/TermsQueryBuilderTests.java (hunk 6 above)
Comment (gf2121) [reply server/src/test/java/org/elasticsearch/index/query/TermsQueryBuilderTests.java]: I removed this method because this is also testing a serialization between different branches, as mentioned [here](https://github.com/elastic/elasticsearch/pull/67223#discussion_r557422972).

File: server/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java [hunk 9]
```diff
@@ -500,4 +399,255 @@ protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) {
...
+        }
+    }
+
+    /**
+     * Store terms as a {@link BytesReference}.
+     * <p>
+     * When users send a query contain a lot of terms, A {@link BytesReference} can help
+     * gc and reduce the cost of {@link #doWriteTo}, which can be slow for lots of terms.
+     */
+    private static class ValuesAfterV8 extends Values {
```
[the remark this answers is not in the document]
Comment (gf2121) [reply server/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java]: I'd like to call them `BinaryValues` and `ListValues`.

**Comment by gf2121** [comment]
@jimczi Thank you very much for the wonderful suggestions! I've done some fix in [24eb238](https://github.com/elastic/elasticsearch/pull/67223/commits/24eb23839d721348e7a80507ca13e8e808410350). Ready for another look:)

**Comment by jimczi** [comment]
@elasticmachine ok to test

File: x-pack/plugin/mapper-unsigned-long/src/main/java/org/elasticsearch/xpack/unsignedlong/UnsignedLongFieldMapper.java [hunk 10]
```diff
@@ -170,7 +171,7 @@ public Query termQuery(Object value, SearchExecutionContext context) {
         }
 
         @Override
-        public Query termsQuery(List<?> values, SearchExecutionContext context) {
+        public Query termsQuery(Collection<?> values, SearchExecutionContext context) {
             failIfNotIndexed();
             long[] lvalues = new long[values.size()];
             int upTo = 0;
```
Comment (jimczi) [inline x-pack/plugin/mapper-unsigned-long/src/main/java/org/elasticsearch/xpack/unsignedlong/UnsignedLongFieldMapper.java:177]: The code below tries to access the values by index so the compilation is broken. Can you fix ?

Comment (gf2121) [reply x-pack/plugin/mapper-unsigned-long/src/main/java/org/elasticsearch/xpack/unsignedlong/UnsignedLongFieldMapper.java:177]: @jimczi I'm so sorry for missing a precommit check! It looks good now.

**Comment by jimczi** [comment]
@elasticmachine ok to test

**Comment by gf2121** [comment]
@jimczi Hi, checks have passed :)

**Review by jimczi — APPROVED** [review]
LGTM, thanks @gf2121 !
I'll merge soon
